File: CDO/outputs/note_uploader/post_to_note.mjs
#!/usr/bin/env node
/**
 * post_to_note.mjs — note.com 完全自動投稿スクリプト
 *
 * 実行：
 *   node post_to_note.mjs <記事ファイル.md>           # フル自動投稿（デフォルト）
 *   node post_to_note.mjs <記事ファイル.md> --review  # 公開直前で止める
 *   node post_to_note.mjs <記事ファイル.md> --dry-run # 入力のみ・公開しない
 *
 * 自動化範囲：
 *   1. 新規投稿ページを開く
 *   2. タイトル入力
 *   3. 本文入力
 *   4. ハッシュタグ追加
 *   5. 「公開設定」クリック
 *   6. 「公開」ボタン押下（--review/--dry-run 指定時はスキップ）
 *
 * 入力ファイル形式（frontmatter）：
 *   ---
 *   title: 記事タイトル
 *   tags:
 *     - Claude
 *     - AI
 *   ---
 *   本文ここから...
 *
 * ⚠ 注意：
 *   - 投稿後に typo 等あれば note 上で編集/削除可能
 *   - noteの利用規約上、自動投稿は明示禁止されていないが、過度な利用はBANリスク
 */

import { chromium } from 'playwright';
import { readFileSync, existsSync } from 'node:fs';
import { dirname, join, resolve } from 'node:path';
import { fileURLToPath } from 'node:url';

const __dirname = dirname(fileURLToPath(import.meta.url));
const AUTH_FILE = join(__dirname, 'auth-state.json');

// ─────────────────────────────────────────────
// frontmatter パーサ
// ─────────────────────────────────────────────
function parseFrontmatter(content) {
  const match = content.match(/^---\r?\n([\s\S]*?)\r?\n---\r?\n([\s\S]*)$/);
  if (!match) return { meta: {}, body: content.trim() };

  const meta = {};
  const lines = match[1].split(/\r?\n/);
  let currentKey = null;

  for (const line of lines) {
    const kv = line.match(/^([a-zA-Z_]+):\s*(.*)$/);
    if (kv) {
      currentKey = kv[1];
      const value = kv[2].trim();
      meta[currentKey] = value === '' ? [] : value.replace(/^["']|["']$/g, '');
    } else if (currentKey && Array.isArray(meta[currentKey])) {
      const item = line.match(/^\s*-\s+(.+)$/);
      if (item) meta[currentKey].push(item[1].trim().replace(/^["']|["']$/g, ''));
    }
  }
  return { meta, body: match[2].trim() };
}

// ─────────────────────────────────────────────
// 引数パース
// ─────────────────────────────────────────────
const args = process.argv.slice(2);
const flags = {
  review:  args.includes('--review'),
  dryRun:  args.includes('--dry-run'),
};
const filePath = args.find((a) => !a.startsWith('--'));

if (!filePath) {
  console.error('使い方: node post_to_note.mjs <記事ファイル.md> [--review | --dry-run]');
  process.exit(1);
}

if (!existsSync(AUTH_FILE)) {
  console.error('❌ auth-state.json が見つかりません。');
  console.error('   先に: npm run login を実行してください。');
  process.exit(1);
}

const absPath = resolve(filePath);
if (!existsSync(absPath)) {
  console.error(`❌ ファイルが見つかりません: ${absPath}`);
  process.exit(1);
}

const content = readFileSync(absPath, 'utf-8');
const { meta, body } = parseFrontmatter(content);

if (!meta.title) {
  console.error('❌ frontmatter に title: が必要です');
  process.exit(1);
}

const tags = Array.isArray(meta.tags) ? meta.tags : [];
const mode = flags.dryRun ? 'DRY-RUN（入力のみ・公開しない）'
           : flags.review ? 'REVIEW（公開直前で停止）'
           : '⚠ FULL-AUTO（公開ボタンまで自動押下）';

console.log('');
console.log('📝 投稿準備');
console.log('================================');
console.log(`タイトル: ${meta.title}`);
console.log(`タグ:     ${tags.join(', ') || '（なし）'}`);
console.log(`本文長:   ${body.length} 文字`);
console.log(`モード:   ${mode}`);
console.log('');

// ─────────────────────────────────────────────
// ヘルパー
// ─────────────────────────────────────────────
async function waitForFirst(page, selectors, timeoutPerSelector = 5000) {
  for (const sel of selectors) {
    try {
      const el = await page.waitForSelector(sel, {
        state: 'visible',
        timeout: timeoutPerSelector,
      });
      if (el) return { el, selector: sel };
    } catch {}
  }
  return null;
}

async function clickByText(page, texts, timeout = 10000) {
  for (const text of texts) {
    try {
      const button = page.getByRole('button', { name: text }).first();
      await button.waitFor({ state: 'visible', timeout });
      await button.click();
      return true;
    } catch {}
  }
  return false;
}

// ─────────────────────────────────────────────
// メイン処理
// ─────────────────────────────────────────────
const browser = await chromium.launch({ headless: false });
const context = await browser.newContext({ storageState: AUTH_FILE });
const page = await context.newPage();

try {
  console.log('🌐 noteの新規投稿ページを開いています...');
  await page.goto('https://note.com/notes/new', { waitUntil: 'domcontentloaded' });
  await page.waitForLoadState('networkidle', { timeout: 30000 }).catch(() => {});

  // goto でホーム等にリダイレクトされた場合の救済：投稿リンクをクリックして編集画面へ
  if (!page.url().includes('/notes/new') && !page.url().includes('/new')) {
    console.log(`ℹ️  URLがリダイレクトされました: ${page.url()}`);
    console.log('   「投稿」リンクを探してクリックします...');
    const newLink = await waitForFirst(page, [
      'a[href*="/notes/new"]',
      'a[href="/new"]',
      'a[href*="editor.note.com"]',
    ], 5000);
    if (newLink) {
      await newLink.el.click();
      await page.waitForLoadState('networkidle', { timeout: 15000 }).catch(() => {});
    }
  }

  // ─── タイトル入力 ───
  console.log('✏️  タイトルを入力中...');
  const titleMatch = await waitForFirst(page, [
    'textarea[placeholder="記事タイトル"]',
    'textarea[placeholder*="タイトル"]',
    'textarea[aria-label*="タイトル"]',
    'input[placeholder*="タイトル"]',
    '[contenteditable="true"][aria-label*="タイトル"]',
  ], 10000);
  if (!titleMatch) {
    throw new Error(`タイトル欄が見つかりません。URL: ${page.url()}`);
  }
  console.log(`   使用セレクタ: ${titleMatch.selector}`);
  await titleMatch.el.click();
  try { await titleMatch.el.fill(meta.title); }
  catch { await page.keyboard.type(meta.title, { delay: 10 }); }

  // ─── 本文入力 ───
  console.log('✏️  本文を入力中...（やや時間がかかります）');
  const bodyMatch = await waitForFirst(page, [
    'div.ProseMirror[contenteditable="true"][role="textbox"]',
    'div[contenteditable="true"].ProseMirror',
    '.ProseMirror[contenteditable="true"]',
    'div[contenteditable="true"][aria-label*="本文"]',
    'div[contenteditable="true"][role="textbox"]:not([aria-label*="タイトル"])',
  ], 10000);
  if (!bodyMatch) {
    throw new Error(`本文欄が見つかりません。URL: ${page.url()}`);
  }
  console.log(`   使用セレクタ: ${bodyMatch.selector}`);
  await bodyMatch.el.click();

  const paragraphs = body.split(/\n\n+/);
  for (let i = 0; i < paragraphs.length; i++) {
    const p = paragraphs[i].replace(/\n/g, ' ').trim();
    if (!p) continue;
    await page.keyboard.type(p, { delay: 2 });
    if (i < paragraphs.length - 1) {
      await page.keyboard.press('Enter');
    }
  }

  await page.waitForTimeout(3000); // 自動保存待機

  if (flags.dryRun) {
    console.log('');
    console.log('✨ DRY-RUN 完了。入力までで停止します。');
    console.log('   ブラウザは開いたままです。Ctrl+C で終了。');
    await new Promise(() => {});
  }

  // ─── 公開設定ページへ進む ───
  console.log('📤 「公開設定」へ進みます...');
  const movedToPublish = await clickByText(page, ['公開設定', '公開に進む']);
  if (!movedToPublish) throw new Error('「公開設定」ボタンが見つかりません');
  await page.waitForLoadState('networkidle', { timeout: 15000 }).catch(() => {});
  await page.waitForTimeout(2000);

  // ─── ハッシュタグ追加 ───
  if (tags.length) {
    console.log(`🏷  ハッシュタグを追加中: ${tags.join(', ')}`);
    const tagInput = await page.$('input[placeholder*="ハッシュタグ"]')
                  || await page.$('input[placeholder*="タグ"]')
                  || await page.$('input[type="text"][aria-label*="タグ"]');
    if (tagInput) {
      for (const tag of tags) {
        await tagInput.click();
        await page.keyboard.type(tag, { delay: 20 });
        await page.keyboard.press('Enter');
        await page.waitForTimeout(500);
      }
    } else {
      console.warn('⚠  タグ入力欄が見つかりません。手動で追加してください。');
    }
  }

  if (flags.review) {
    console.log('');
    console.log('✨ REVIEW モード：公開直前で停止します。');
    console.log('   内容を確認してから手動で「公開」ボタンを押してください。');
    console.log('   ブラウザはこのまま。Ctrl+C で終了。');
    await new Promise(() => {});
  }

  // ─── 公開ボタンクリック ───
  console.log('🚀 公開ボタンを押下中...');
  const published = await clickByText(page, ['投稿する', '公開する', '公開', '有料エリアありで公開']);
  if (!published) throw new Error('「公開」ボタンが見つかりません');

  await page.waitForTimeout(5000);

  // ─── 公開後URL取得 ───
  const finalUrl = page.url();
  console.log('');
  console.log('==================================================');
  console.log('✅ 公開完了');
  console.log('==================================================');
  console.log(`URL: ${finalUrl}`);
  console.log('');
  console.log('5秒後にブラウザを閉じます...');
  await page.waitForTimeout(5000);
  await browser.close();
  process.exit(0);

} catch (err) {
  console.error('');
  console.error('❌ エラー:', err.message);
  const shotPath = join(__dirname, `error_${Date.now()}.png`);
  await page.screenshot({ path: shotPath, fullPage: true }).catch(() => {});
  console.error(`スクショ保存: ${shotPath}`);
  console.error('ブラウザは開いたままです。手動で続行するか Ctrl+C で終了。');
  await new Promise(() => {});
}

File: projects/2026-05-20_note記事毎日自動投稿/scripts/post_to_note.py
"""
post_to_note.py — note.com への自動投稿スクリプト

実行環境：
- GitHub Actions（推奨。Secretsから認証情報を読む）
- ローカル（.envに認証情報を置く）

必要な環境変数：
- NOTE_EMAIL              ログイン用メールアドレス
- NOTE_PASSWORD           ログイン用パスワード
- NOTE_SESSION_COOKIE     （任意）事前取得したセッションcookie JSON
- ARTICLE_PATH            （任意）投稿する記事のリポジトリ内パス
- DRY_RUN                 "true"ならログイン確認のみで投稿せず終了
"""

from __future__ import annotations

import json
import os
import re
import sys
from datetime import datetime
from pathlib import Path

REPO_ROOT = Path(__file__).resolve().parents[3]
PROJECT_DIR = Path(__file__).resolve().parents[1]
STATE_DIR = PROJECT_DIR / "state"
STATE_FILE = STATE_DIR / "posted.json"
DEBUG_DIR = Path(__file__).parent

NOTE_LOGIN_URL = "https://note.com/login"
NOTE_NEW_URL = "https://note.com/notes/new"


def load_posted_state() -> dict:
    if not STATE_FILE.exists():
        return {"posted": []}
    return json.loads(STATE_FILE.read_text(encoding="utf-8"))


def save_posted_state(state: dict) -> None:
    STATE_DIR.mkdir(parents=True, exist_ok=True)
    STATE_FILE.write_text(
        json.dumps(state, ensure_ascii=False, indent=2),
        encoding="utf-8",
    )


def pick_article_to_post(explicit_path: str | None) -> Path:
    """投稿する記事のパスを決定する。

    explicit_path が指定されていればそれを使う。
    なければ CMO/outputs/ 配下の最新 *_note記事.md で、未投稿のものを選ぶ。
    """
    if explicit_path:
        p = (REPO_ROOT / explicit_path).resolve()
        if not p.exists():
            raise FileNotFoundError(f"指定された記事が見つかりません: {p}")
        return p

    outputs_dir = REPO_ROOT / "CMO" / "outputs"
    candidates = sorted(outputs_dir.glob("*_note記事.md"), reverse=True)
    posted_paths = set(load_posted_state().get("posted", []))
    for c in candidates:
        rel = str(c.relative_to(REPO_ROOT))
        if rel not in posted_paths:
            return c
    raise RuntimeError("投稿可能な未投稿記事が見つかりません")


def parse_article(article_path: Path) -> tuple[str, str]:
    """Markdownの記事から、タイトルと本文を抽出する。

    H1（# タイトル）をタイトルにし、それ以外を本文とする。
    """
    text = article_path.read_text(encoding="utf-8")
    lines = text.splitlines()
    title = None
    body_lines: list[str] = []
    for line in lines:
        if title is None and line.startswith("# "):
            title = line[2:].strip()
            continue
        body_lines.append(line)
    if not title:
        title = article_path.stem
    body = "\n".join(body_lines).strip()
    return title, body


def normalize_cookie(c: dict) -> dict | None:
    """Chrome拡張「EditThisCookie」やDevToolsからエクスポートしたcookieを
    playwright の add_cookies が受け取れる形に正規化する。
    """
    if "name" not in c or "value" not in c:
        return None
    out: dict = {
        "name": c["name"],
        "value": c["value"],
        "path": c.get("path", "/"),
    }
    if "domain" in c and c["domain"]:
        out["domain"] = c["domain"]
    elif "url" in c:
        out["url"] = c["url"]
    else:
        out["domain"] = ".note.com"
    if "expirationDate" in c:
        try:
            out["expires"] = float(c["expirationDate"])
        except (TypeError, ValueError):
            pass
    elif "expires" in c:
        try:
            out["expires"] = float(c["expires"])
        except (TypeError, ValueError):
            pass
    if "httpOnly" in c:
        out["httpOnly"] = bool(c["httpOnly"])
    if "secure" in c:
        out["secure"] = bool(c["secure"])
    if "sameSite" in c and c["sameSite"]:
        ss = str(c["sameSite"]).strip()
        ss_lower = ss.lower()
        if ss_lower in ("strict", "lax", "none"):
            out["sameSite"] = ss.capitalize() if ss_lower != "none" else "None"
        elif ss_lower in ("no_restriction", "unspecified"):
            out["sameSite"] = "None"
    return out


def login_with_cookie(context, cookie_json: str) -> bool:
    """事前取得済みのセッションcookieでログイン状態を再現する。"""
    try:
        raw = json.loads(cookie_json)
    except json.JSONDecodeError:
        print("[WARN] NOTE_SESSION_COOKIE がJSON parseできません。パスワードログインにフォールバックします。")
        return False
    if not isinstance(raw, list):
        print("[WARN] NOTE_SESSION_COOKIE は配列形式である必要があります")
        return False
    normalized = [c for c in (normalize_cookie(c) for c in raw) if c]
    if not normalized:
        print("[WARN] 有効なcookieが0件です")
        return False
    try:
        context.add_cookies(normalized)
    except Exception as e:
        print(f"[WARN] cookie注入に失敗: {e}")
        return False
    print(f"[INFO] {len(normalized)}件のcookieを注入しました")
    return True


def login_with_password(page, email: str, password: str) -> None:
    """メール+パスワードでログインする。

    note.comのログインフォームは将来UIが変わる可能性があるため、
    複数のセレクタ候補を試す。
    """
    page.goto(NOTE_LOGIN_URL, wait_until="domcontentloaded")
    page.wait_for_timeout(2000)

    email_selectors = [
        'input[type="email"]',
        'input[name="email"]',
        'input[placeholder*="メール"]',
    ]
    password_selectors = [
        'input[type="password"]',
        'input[name="password"]',
    ]

    for sel in email_selectors:
        if page.locator(sel).count():
            page.locator(sel).first.fill(email)
            break
    else:
        page.screenshot(path=str(DEBUG_DIR / "debug_login_email_not_found.png"))
        raise RuntimeError("メール入力欄が見つかりません（note.comのUI変更の可能性）")

    for sel in password_selectors:
        if page.locator(sel).count():
            page.locator(sel).first.fill(password)
            break
    else:
        page.screenshot(path=str(DEBUG_DIR / "debug_login_password_not_found.png"))
        raise RuntimeError("パスワード入力欄が見つかりません")

    submit_selectors = [
        'button[type="submit"]',
        'button:has-text("ログイン")',
    ]
    for sel in submit_selectors:
        if page.locator(sel).count():
            page.locator(sel).first.click()
            break

    try:
        page.wait_for_url(re.compile(r"https://note\.com/(?!login).*"), timeout=15000)
    except Exception:
        page.screenshot(path=str(DEBUG_DIR / "debug_login_no_redirect.png"))
        raise RuntimeError("ログイン後のリダイレクトが確認できません（2FA/CAPTCHAの可能性）")


def post_article(page, title: str, body: str) -> str:
    """noteの新規投稿画面で記事を入力して公開する。

    noteのエディタはリッチテキストで、Markdownをそのままペーストすると
    一部書式が崩れる可能性がある。当面はプレーンテキストとして投入し、
    UI構造が固まったら段階的に整形を強化する。
    """
    page.goto(NOTE_NEW_URL, wait_until="domcontentloaded")
    page.wait_for_timeout(3000)

    title_selectors = [
        'textarea[placeholder*="タイトル"]',
        'input[placeholder*="タイトル"]',
        '[contenteditable="true"][data-placeholder*="タイトル"]',
    ]
    for sel in title_selectors:
        if page.locator(sel).count():
            page.locator(sel).first.fill(title)
            break
    else:
        page.screenshot(path=str(DEBUG_DIR / "debug_editor_title_not_found.png"))
        raise RuntimeError("タイトル入力欄が見つかりません")

    body_selectors = [
        '[contenteditable="true"][data-placeholder*="本文"]',
        'div.ProseMirror',
        '[role="textbox"]',
    ]
    body_locator = None
    for sel in body_selectors:
        if page.locator(sel).count():
            body_locator = page.locator(sel).last
            break
    if body_locator is None:
        page.screenshot(path=str(DEBUG_DIR / "debug_editor_body_not_found.png"))
        raise RuntimeError("本文入力欄が見つかりません")

    body_locator.click()
    body_locator.type(body, delay=5)

    page.wait_for_timeout(2000)
    page.screenshot(path=str(DEBUG_DIR / "debug_before_publish.png"))

    publish_selectors = [
        'button:has-text("公開設定")',
        'button:has-text("公開に進む")',
        'button:has-text("投稿")',
    ]
    for sel in publish_selectors:
        if page.locator(sel).count():
            page.locator(sel).first.click()
            page.wait_for_timeout(2000)
            break

    final_publish_selectors = [
        'button:has-text("公開する")',
        'button:has-text("有料公開")',
    ]
    for sel in final_publish_selectors:
        if page.locator(sel).count():
            page.locator(sel).first.click()
            break

    try:
        page.wait_for_url(re.compile(r"https://note\.com/.+/n/.+"), timeout=20000)
    except Exception:
        page.screenshot(path=str(DEBUG_DIR / "debug_publish_no_redirect.png"))
        raise RuntimeError("公開後のURLが取得できません")

    return page.url


def main() -> int:
    email = os.environ.get("NOTE_EMAIL", "").strip()
    password = os.environ.get("NOTE_PASSWORD", "").strip()
    cookie_json = os.environ.get("NOTE_SESSION_COOKIE", "").strip()
    explicit = os.environ.get("ARTICLE_PATH", "").strip()
    dry_run = os.environ.get("DRY_RUN", "").lower() == "true"

    if not (email and password) and not cookie_json:
        print("[ERROR] 認証情報が設定されていません。")
        print("  GitHub Secrets に NOTE_EMAIL+NOTE_PASSWORD または NOTE_SESSION_COOKIE を設定してください。")
        return 2

    article_path = pick_article_to_post(explicit or None)
    title, body = parse_article(article_path)
    print(f"[INFO] 投稿対象: {article_path.relative_to(REPO_ROOT)}")
    print(f"[INFO] タイトル: {title}")
    print(f"[INFO] 本文長: {len(body)} 文字")

    try:
        from playwright.sync_api import sync_playwright
    except ImportError:
        print("[ERROR] playwright が未インストールです。`pip install playwright && playwright install chromium`")
        return 3

    with sync_playwright() as p:
        browser = p.chromium.launch(headless=True)
        context = browser.new_context(
            viewport={"width": 1280, "height": 900},
            user_agent=(
                "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 "
                "(KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36"
            ),
        )
        page = context.new_page()

        logged_in = False
        if cookie_json:
            logged_in = login_with_cookie(context, cookie_json)
            if logged_in:
                # 投稿画面に直接遷移してリダイレクトされなければログイン済み
                try:
                    page.goto(NOTE_NEW_URL, wait_until="domcontentloaded", timeout=15000)
                    page.wait_for_timeout(2500)
                except Exception as e:
                    print(f"[WARN] 投稿画面遷移エラー: {e}")
                    page.screenshot(path=str(DEBUG_DIR / "debug_cookie_goto_error.png"))
                    logged_in = False

                if logged_in:
                    current_url = page.url
                    print(f"[INFO] After goto, URL = {current_url}")
                    page.screenshot(path=str(DEBUG_DIR / "debug_after_cookie_login.png"))
                    try:
                        page.content_html = page.content()
                        with open(DEBUG_DIR / "debug_after_cookie_login.html", "w", encoding="utf-8") as f:
                            f.write(page.content_html[:50000])
                    except Exception:
                        pass

                    # ログイン画面に飛ばされていたら失敗
                    if "/login" in current_url or "/signin" in current_url:
                        print("[WARN] 投稿画面アクセス時にログイン画面へリダイレクトされました（cookie失効か無効）")
                        logged_in = False
                    else:
                        print("[INFO] cookieログイン成功（投稿画面に直接アクセスできました）")

        if not logged_in:
            if not (email and password):
                print("[ERROR] cookieログインに失敗、パスワードも未設定です")
                return 4
            login_with_password(page, email, password)

        if dry_run:
            print("[INFO] DRY_RUN: ログイン確認OK、投稿はスキップします")
            return 0

        url = post_article(page, title, body)
        print(f"[OK] 投稿完了: {url}")

        state = load_posted_state()
        state.setdefault("posted", []).append(str(article_path.relative_to(REPO_ROOT)))
        state.setdefault("history", []).append(
            {
                "path": str(article_path.relative_to(REPO_ROOT)),
                "url": url,
                "posted_at": datetime.utcnow().isoformat() + "Z",
            }
        )
        save_posted_state(state)
        browser.close()
    return 0


if __name__ == "__main__":
    sys.exit(main())
